testsupport/wait/awaitility.go

```diff
@@ -658,3 +662,89 @@ func (a *Awaitility) listAndReturnContent(resourceKind, namespace string, list c
 	content, _ := StringifyObjects(list)
 	return fmt.Sprintf("\n%s present in the namespace:\n%s\n", resourceKind, string(content))
 }
+
+// Waiter is a helper struct for `wait.For()` that provides functions to query the cluster waiting
+// for the results.
+type Waiter[T client.Object] struct {
+	await *Awaitility
+	t     *testing.T
+	gvk   schema.GroupVersionKind
+}
+
+// AtLeast merely waits for the given amount of time ensuring that "wait.For" (from which it is called) is
+// happening at least the duration needed.
+func (w *Waiter[T]) AtLeast(dur time.Duration) *Waiter[T] {
+	time.Sleep(dur)
+	return w
+}
+
+// FirstThat uses the provided predicates to filter the objects of the type provided to `wait.For()` and
+// repeatedly tries to find the first one that satisfies all the predicates.
+func (w *Waiter[T]) FirstThat(predicates ...predicates.Predicate[client.Object]) (T, bool, error) {
+	w.t.Logf("waiting for objects of GVK '%s' in namespace '%s' to match criteria", w.gvk, w.await.Namespace)
+
+	var returnedObject T
+	found := false
+
+	err := wait.Poll(w.await.RetryInterval, w.await.Timeout, func() (done bool, err error) {
+		// because there is no generic way of figuring out the list type for some client.Object type, we need to go
+		// down the low level route and use unstructured to get the list generically and unmarshal and cast the list
+		// items.
+		list := &unstructured.UnstructuredList{}
+		list.SetGroupVersionKind(w.gvk)
+		if err := w.await.Client.List(context.TODO(), list, client.InNamespace(w.await.Namespace)); err != nil {
+			return false, err
+		}
+		for _, obj := range list.Items {
+			raw, err := obj.MarshalJSON()
+			if err != nil {
+				return false, fmt.Errorf("failed to obtain the raw JSON of object when getting objects of type %v: %w", w.gvk, err)
+			}
+
+			typed, err := w.await.Client.Scheme().New(w.gvk)
+			if err != nil {
+				return false, fmt.Errorf("failed to create a new empty object of type %v: %w", w.gvk, err)
+			}
+
+			err = json.Unmarshal(raw, typed)
+			if err != nil {
+				return false, fmt.Errorf("failed to unmarshal the raw JSON to the go structure of object type: %v: %w", w.gvk, err)
+			}
+
+			object := typed.(T)
+
+			matches := true
+			for _, p := range predicates {
+				if !p.Matches(object) {
+					matches = false
+					break
+				}
+			}
+
+			if matches {
+				found = true
+				returnedObject = object
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	return returnedObject, found, err
```

Comment:
I wanted it to be easier for the user to determine whether the call found something or failed with some kind of other error. Of course, you are right that the caller could figure that out from the error itself by checking:

```go
obj, err := wait.For(t, a, &corev1.Pod{}).FirstThat(Runs())
assert.False(t, errors.Is(err, kwait.ErrWaitTimeout))
```

but I think it is more ergonomic to just use the bool from the result (or ignore it for that matter):

```go
obj, found, err := wait.For(t, a, &corev1.Pod{}).FirstThat(Runs())
assert.True(t, found)
```

Reply:
following the best practice, in 99% of all cases, the code handles the error first
```
obj, found, err := wait.For(t, a, &corev1.Pod{}).FirstThat(Runs())
require.NoError(t, err)
```
and if the function returns an error in case the object wasn't found, then there is not much of a point in having the extra returned value boolean `found` - it doesn't bring some additional value.

Reply:
also, in this particular cases - it's the assumption/expectations/assertion which failed, so it's totally fine to rely on the error.
If we would need to provide some additional information needed for the following execution of the test, then let's add the extra returned value, otherwise, let's keep it simple and rely on the error only.